compiler/src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -1153,7 +1153,12 @@ class Namer { typer: Typer =>
       case _ =>
         WildcardType
     }
-    paramFn(typedAheadType(mdef.tpt, tptProto).tpe)
+    val tp = typedAheadType(mdef.tpt, tptProto).tpe
+    paramFn(
+      if (tp.isValueType || (tp match {
+        case ExprType(tp1) => tp1.isValueType
+        case _ => false
+      })) tp else ErrorType("A type of a value or a return type of a method should be a value type."))
```

Comment:
It's more succinct to write `tp.widenExpr.isValueType` here. I'm curious which test fails without the widening?

Reply:
I'll fix it. Thank you!
When a method has a by-name parameter, typing `ValDef` of the parameter results `ExprType`.